### ISSUE TITLE
About BURST

### ISSUE THREAD

**Issue by z-jiaming** [hkchengrex/Cutie#4]
Hi!

First of all, thanks for your great work and code, specifically the new Dataloader and BURST tools.
But, I have some questions about the following things:

1) Code issue
convert_burst_to_vos_train.py seems to have several problems

- remove # L28
- #L65: output_folder should be visualize_folder

2) BURST Tools
If my understanding is correct, I could use the convert_burst_to_vos_train.py to convert burst mask in json to png.

- `python scripts/convert_burst_to_vos_train.py --json_path ../BURST/annotations/train/train.json --input_path ../BURST/frames/train --output_path ../BURST/vos_train/train # generate train split, which have all masks`
- `python scripts/convert_burst_to_vos_train.py --json_path ../BURST/annotations/val/first_frame_annotations.json --input_path ../BURST/frames/val --output_path ../BURST/vos_train/val_first_annotations # generate val split, which have the first annotations of each object`
- `python scripts/convert_burst_to_vos_train.py --json_path ../BURST/annotations/val/all_classes.json --input_path ../BURST/frames/val --output_path ../BURST/vos_train/val_all_annotations # generate val split, which have all annotations of each object`

I use the all_classes.json to generate full mask for val split, then use mask_to_burst_json.py to re-convert it to predition json.
I eval the predition json, but the metric it not 100 for both val and test split.

<img width="654" alt="image" src="https://github.com/hkchengrex/Cutie/assets/42411674/d2bfe0d2-1aaa-4876-b0fd-ee31345cc9aa">

If these two tools are absolutely correct and there is no loss, i.e. json->png and png->json, why the metric is not 100?

3) Your precomputed_results

I downloaded your results (cutie-base-wmose_burst-val.zip) to verify the correctness of my evaluation tool for BURST-benchmark.
Although the metric is the same as yours, when I use mask_to_burst_json.py, I find all the pngs in Annotations are null like this.

<img width="1183" alt="image" src="https://github.com/hkchengrex/Cutie/assets/42411674/36c54b76-c607-4daf-b54b-91320d5662da">

which can not use mask_to_burst_json.py, I want to know the reason for this.

Sorry for the many questions, thank you for your patient reply!

**Comment by hkchengrex** [comment]
1. RIght. I will fix it. Thanks.
2 and 3. I haven't tried testing it cyclically -- let me investigate. I'll get back to you.

**Comment by hkchengrex** [comment]
I find that BURST annotation contains some overlapping masks. Information loss is thus inevitable when converting them to the mask format. Given that all the scores are very high (99+), the loss should be minimal.

I cannot reproduce the empty mask problem though.

**Comment by z-jiaming** [comment]
Thank you for your patient reply!

For the empty mask problem, I compared your provided results with converted annotations using part codes of mask_to_burst_json.py as follows:
```
import sys
import os
import json
from collections import defaultdict

from PIL import Image
import numpy as np
import pycocotools.mask as mask_utils
import tqdm


cutie_path = "Cutie/github/cutie-base-wmose_burst-val/Annotations/00c561b9-2057-358d-82c6-5b06d76cebcf/ring_front_center_315969629022515560.png"
anno_path = "BURST/val/Annotations_all_classes/ArgoVerse_-_00c561b9-2057-358d-82c6-5b06d76cebcf/ring_front_center_315969629022515560.png" # use all_classes.json

if os.path.exists(cutie_path):
    cutie_mask = np.array(Image.open(cutie_path))
    cutie_labels = np.unique(cutie_mask)
    cutie_labels = cutie_labels[cutie_labels != 0].tolist()

    for cutie_label in cutie_labels:
        cutie_this_mask = (cutie_mask == cutie_label).astype(np.uint8)
        if cutie_this_mask.sum() == 0:
            continue

        cutie_this_mask = mask_utils.encode(np.asfortranarray(cutie_this_mask))
        
if os.path.exists(anno_path):
    anno_mask = np.array(Image.open(anno_path))
    anno_labels = np.unique(anno_mask)
    anno_labels = anno_labels[anno_labels != 0].tolist()

    for anno_label in anno_labels:
        anno_this_mask = (anno_mask == anno_label).astype(np.uint8)
        if anno_this_mask.sum() == 0:
            continue

        anno_this_mask = mask_utils.encode(np.asfortranarray(anno_this_mask))

print()
```

You can see anno_this_mask has a different format with cutie_this_mask, which can not use
```
this_mask = mask_utils.encode(np.asfortranarray(this_mask))
this_mask['counts'] = this_mask['counts'].decode('utf-8')
this_segment[label] = {'rle': this_mask['counts']}
```

<img width="1519" alt="image" src="https://github.com/hkchengrex/Cutie/assets/42411674/60e35625-b483-4a50-818a-dbe486fe599b">


These are your provided results and my converted annotations:
![CUTIE_ring_front_center_315969629022515560](https://github.com/hkchengrex/Cutie/assets/42411674/6ffeb497-0ea9-4458-a1fb-507622606c45)
![MY_ring_front_center_315969629022515560](https://github.com/hkchengrex/Cutie/assets/42411674/d97afcb4-3312-4275-92bf-28ae34ba0a20)

**Comment by z-jiaming** [comment]
Again, thank you a lot for your patient reply!

1) I just figure out what the difference is between these two masks.
As with the first image(cutie-base-wmose_burst-val/Annotations/00c561b9-2057-358d-82c6-5b06d76cebcf/ring_front_center_315969629022515560.png), it is all black and it is impossible to visualize the results you are providing.
Or do you have any way to convert it to a viewable result?

2) I use convert_burst_to_vos_train.py to convert val and test split, then re-run eval_vos.py with the converted dataset.
I modify the vos_test_dataset.py like:
```
import os
from os import path
import json
from typing import Iterable, Optional

from cutie.inference.data.video_reader import VideoReader


class VOSTestDataset:
    def __init__(self,
                 image_dir: str,
                 mask_dir: str,
                 *,
                 use_all_masks: bool,
                 req_frames_json: Optional[str] = None,
                 size: int = -1,
                 size_dir: Optional[str] = None,
                 subset: Optional[str] = None):
        self.image_dir = image_dir
        self.mask_dir = mask_dir
        self.use_all_masks = use_all_masks
        self.size = size
        self.size_dir = size_dir

        if subset is None:
            self.vid_list = sorted(os.listdir(self.mask_dir))
        else:
            # DAVIS-2017 style txt
            with open(subset) as f:
                self.vid_list = sorted([line.strip() for line in f])

        self.req_frame_list = {}
        if req_frames_json is not None:
            if 'BURST' in image_dir:
                with open(req_frames_json) as f:
                    meta = json.load(f)['sequences']

                    for vid in meta:
                        assert (vid['dataset'] + '_-_' + vid['seq_name']) in self.vid_list, '{} not in {}'.format((vid['dataset'] + '_-_' + vid['seq_name']), self.vid_list)
                        assert sorted(os.listdir(os.path.join(self.image_dir, (vid['dataset'] + '_-_' + vid['seq_name'])))) == vid['all_image_paths']

                        self.req_frame_list[(vid['dataset'] + '_-_' + vid['seq_name'])] = [img[:-4] for img in vid['annotated_image_paths']]
            elif 'YouTubeVOS' in image_dir: # YouTubeVOS style json
                with open(req_frames_json) as f:
                    # read meta.json to know which frame is required for evaluation
                    meta = json.load(f)['videos']

                    for vid in self.vid_list:
                        req_frames = []
                        objects = meta[vid]['objects']
                        for value in objects.values():
                            req_frames.extend(value['frames'])

                        req_frames = list(set(req_frames))
                        self.req_frame_list[vid] = req_frames
            else:
                raise NotImplementedError

    def get_datasets(self) -> Iterable[VideoReader]:
        for video in self.vid_list:
            yield VideoReader(
                video,
                path.join(self.image_dir, video),
                path.join(self.mask_dir, video),
                size=self.size,
                to_save=self.req_frame_list.get(video, None),
                use_all_masks=self.use_all_masks,
                size_dir=path.join(self.size_dir, video) if self.size_dir is not None else None,
            )

    def __len__(self):
        return len(self.vid_list)

```

But I get inconsistent results, using the generic VOSTestDataset will have lower metrics than the BURSTTestDataset.
<img width="1331" alt="image" src="https://github.com/hkchengrex/Cutie/assets/42411674/6b580284-f8dc-4fe8-9592-605053a3f633">
Val and test split are on the left and right, and BURSTestDataset and VOSTestDataset are on the top and bottom, respectively.

It would greatly reduce the difficulty of reviewing BURST if the cause of the inconsistency could be identified, much appreciated!

**Comment by hkchengrex** [comment]
If I use `convert_burst_to_vos_train.py` on the provided precomputed result:
```bash
python scripts/convert_burst_to_vos_train.py --json_path ../precomputed_results/cutie-base-wmose/burst-val/predictions.json --frames_path ../BURST/frames/val --output_path ../BURST/val-vos-precomputed
```

I get a non-empty mask:
![ring_front_center_315969629022515560](https://github.com/hkchengrex/Cutie/assets/7107196/561095cf-f73b-4058-b766-d1f02417a08f)

Is this the step where we get different results?

**Comment by z-jiaming** [comment]
Perhaps I didn't make myself clear, I have no doubt that using `convert_burst_to_vos_train.py` can successfully convert json to mask.
It's just that the save_result() of ResultSaver, would it not be more convenient if BURST was specially designed so that it saves the observable results directly, rather than needing to go through a secondary conversion? 
Just a small suggestion:blush:

**Comment by z-jiaming** [comment]
In addition, for the convenience of evaluating BURST, can `convert_burst_to_vos_train.py` also be used to convert its val and test? If so, it should generate JPEGImages_all_frames as well.

I pull a request on: https://github.com/hkchengrex/Cutie/pull/9

**Comment by hkchengrex** [comment]
I think I got you now. Are you referring to the "Annotations" folder in the provided results? I just noticed that they are not visualized correctly. That is indeed a bug and I have pushed a fix. Thank you for letting me know. I am probably not updating the provided results though because they take quite some time/GPU to generate.
